### ISSUE TITLE
FLIP: Scope Flags

### ISSUE THREAD

**Issue by cgarciae** [google/flax#2131]
# Scope Flags
I'll leave a basic idea here before creating the full FLIP.

This FLIP wants to propose a new mechanism called "flags" that enables a module to have access to some context information that might alter its behaviour, relieving the user from having to repeatedly propagate information down the call stack and enabling modules to get additional information that is not available to them in the current implementation.

The proposal would add a new `flags` keyword argument to `init` and `apply` that would accept a `Dict[str, Any]` value with information that would be propagated through the scope:
```python
module.apply(..., flags={'initializing': False, 'deterministic': True})
```
Module would have 2 new methods, `get_flag` and `has_flag` that would enable it to get access to flag values:
```python
initializing = self.get_flag('initializing') if self.has_flag('initializing') else False
deterministic = self.get_flag('deterministic', True) # default
```
Frameworks like Pytorch or Keras have OO-based mechanisms that serve for similar purposes, e.g. Keras layers have a `.built` and `.training` attributes, Pytorch modules have a `.training` property and `train` + `eval` methods to recursively set the training state. Flags are a more general version of these mechanisms that is compatible with Functional Programming.

## Use Case 1: Module Initialization
Flax doesn't have a special initialization mechanism as `init` under the hood it just `apply` with empty `variables`. Many Modules have been relying on either existence (or not) of certain variables, or the mutable state of a collection. As discussed in #652, `BatchNorm` defines its initialization state as:

```python
initializing = self.is_mutable_collection('params')
```
This is problematic because if a user tries to train with `module.apply(..., mutable=True)` from an outside perspective it is not immediately evident that batch statistics will never be updated since the module believes it is always initializing. A very similar problem is present in the RNN module that is being developed in #2126 as the layer uses the same mechanism -out of lack a better option- to determine if its initializing (in which case it won't update the `carry`). 

A possible solution would be for `init` to automatically add an `'initializing': True` flag that could give a hint to these layers that the user intends to initialize variables. As pointed out by @jheek in #652 this flag might not be a universal solution but in combination with other strategies it can avoid a certain class of issues.

## Use Case 2: Training state and Sequential
Modules that use `Dropout` and `BatchNorm` usually accept a `training: bool` parameter to set `deterministic` and `use_running_average` according to the situation:

```python
class ConvBlock(nn.Module):
    @nn.compact
    def __call__(self, x: jnp.ndarray, training: bool) -> jnp.ndarray:
        x = nn.Conv(32, [3, 3], strides=[2, 2])(x)
        x = nn.BatchNorm(use_running_average=not training)(x)
        x = nn.Dropout(0.2, deterministic=not training)(x)
        x = nn.relu(x)
        return x
```
However, this does not play well with combinators like `Sequential` where this information is not available / cannot be defined a prior:
```python
# this is not easy to do right now
module = nn.Sequential([
    nn.Conv(32, [3, 3], strides=[2, 2]),
    nn.BatchNorm(use_running_average=???),
    nn.Dropout(0.2, deterministic=???),
    nn.relu,
])
```
A possible solution would be for these layers to listen to a single `training` flag or two separate `deterministic` and `use_running_average` flags, such that modules can default to the values of these flags when its not directly provided such that you are able to define the previous like this:
```python
module = nn.Sequential([
    nn.Conv(32, [3, 3], strides=[2, 2]),
    nn.BatchNorm(),
    nn.Dropout(0.2),
    nn.relu,
])

... = module.apply(..., flags={'deterministic': False, 'use_running_average': False})
```

**Comment by andsteing** [comment]
What exactly is this FLIP about? Thanks for adding a description & a priority.

**Comment by cgarciae** [comment]
@andsteing I extended a bit but this serves as a proto FLIP while I port it and add more details.

**Comment by marcvanzee** [comment]
Thanks a lot for writing this FLIP Cristian! 👍 

Here's my 2 cents...

Re: use case 1
it seems to me the scope flags are slightly overkill, and I would prefer to just implement the `init` flag. I think we should not introduce complexity unnecessarily, and as long as we don't have a real need for more flags I would prefer to only add one. If it turns out we need more then we can always decide to implement a more general solution.

Re: use case 2
Is this a real example you encountered? I wonder how often it actually happens that someone uses a `Sequential` which requires setting multiple flags... again I am in favor of only introducing complexity if needed, so I would prefer to wait until we hear this is a problem before finding a solution. For use case 1 we know this is a problem, but I am not sure about this case.

**Comment by jheek** [comment]
> it seems to me the scope flags are slightly overkill, and I would prefer to just implement the init flag

I agree we shouldn't directly expose this but I think it's worth thinking about a generalization regardless. This also helps think through edge cases like meta learning, hyper nets, and cache init where the initialization mode is not clearly a single separate stage

I also like the proposal that @levskaya mentioned about potentially separating initializable and mutable collections. This distinction is currently also missing in for example nn.scan where you can mutate a carried variable collection but you cannot actually initialize anything

**Comment by andsteing** [comment]
wrt use case 2
since module creation is very cheap could you simply?

```python
get_module = lambda train: nn.Sequential([
    nn.Conv(32, [3, 3], strides=[2, 2]),
    nn.BatchNorm(use_running_average=not train),
    nn.Dropout(0.2, deterministic=not train),
    nn.relu,
])
```

**Comment by marcvanzee** [comment]
@cgarciae I think @levskaya implemented a simple version of scopes a while ago. Is this FLIP still relevant?

**Comment by cgarciae** [comment]
Indeed, let's close this.
